tutorials/python/python-type-driven-development/tutorial.md

```diff
@@ -0,0 +1,207 @@
+
+```python
+from typing import Callable, Optional
+
+
+def sum_and_process(a: int, b: int, callback: Callable[[int, Optional[str]], bool]) -> bool:
+   total = a + b
+   return callback(total, None)
+
+
+def is_positive(val: int, message: Optional[str]) -> bool:
+   if message:
+       print(message)
+   return val > 0
+
+
+output = sum_and_process(5, 2, is_positive)
+print(output)
+# True
+```
+
+First of all, we are creating a function that takes two integer values, sums up them and returns the output of an incoming callback. This callback `is_positive` is defined as a function that takes one mandatory integer and one optional string. If the message exists it will print it, and always returns if the incoming value is positive. Of course, this function is not pure because of the print side-effect, but it’s just an example :)
+
+## Generics and Union Types
+
+One of the most powerful features in a type system is the generics and union types. It’s also possible to use these in Python annotations. Let’s take a look at the following example:
+
+```python
+from typing import TypeVar, List
+
+
+T = TypeVar("T", int, List[str])
+
+def generic_add(x: T, y: T) -> T:
+   return x + y
+
+
+x1: int = 5
+y1: int = 2
+print(generic_add(x1, y1))
+# 7
+
+x2: List[str] = ["Hello"]
+y2: List[str] = ["World"]
+print(generic_add(x2, y2))
+# ['Hello', 'World']
+
+x3: str = "foo"
+y3: str = "bar"
+print(generic_add(x3, y3))
+# error: Value of type variable "T" of "generic_add" cannot be "str"
+```
+
+Above, we defined a type variable `T` that can be integer or List of strings and a `generic_add` function that perform the “addition” operation for the incoming arguments which are limited by the `TypeVar` declaration.
+
+Also, we can make use of union types. Let’s suppose we want to support any number that can only be of types integer and float:
+
+```python
+from typing import Union
+
+
+Number = Union[float, int]
+
+def union_add(x: Number, y: Number) -> Number:
+   return x + y
+
+
+x1: int = 5
+y1: int = 2
+print(union_add(x1, y1))
+# 7
+
+x2: float = 3.14
+y2: float = 3.14
+print(union_add(x2, y2))
+# 6.28
+
+x3: str = "2"
+y3: str = "1"
+print(union_add(x3, y3))
+# error: Argument 1 to "union_add" has incompatible type "str"; expected "Union[float, int]"
+# error: Argument 2 to "union_add" has incompatible type "str"; expected "Union[float, int]"
+```
+
+To wrap up, static type checking will always be important to prevent bugs or implementation misuse, and allows us to take one of the main advantages of _Type Driven Development_, don't rely too much in unit tests that check the values consistency like other dynamic and weak languages do. Now you can start doing type checking automatically in your local environments or as a continuous integration job along with the rest of your pipeline.
```

Comment:
Perhaps something nice here would be saying more about the "don't rely too much" part, that is, it's not about one or the other, but both and in this case adding this to Python can improve things so that unit tests can be very useful for something specific, and the new things here can be useful for something else that complements the other one.

Reply:
Section expanded.